crypto/x509/v3_ncons.c

```diff
@@ -581,6 +595,52 @@ static int nc_dns(ASN1_IA5STRING *dns, ASN1_IA5STRING *base)
 
 }
 
+/*
+ * This function implements comparison between ASCII/U-label in eml
+ * and A-label in base according to RFC 8398, section 6.
+ * Convert base to U-label and ASCII-parts of domain names, for base
+ * Octet-to-octet comparison of `eml` and `base` hostname parts
+ * (ASCII-parts should be compared in case-insensitive manner)
+ */
+static int nc_email_eai(ASN1_UTF8STRING *eml, ASN1_IA5STRING *base)
+{
+    const char *baseptr = (char *)base->data;
+    const char *emlptr = (char *)eml->data;
+    const char *emlat = strchr(emlptr, '@');
+
+    char ulabel[256];
+    size_t size = sizeof(ulabel) - 1;
+
+    if (emlat == NULL)
+        return X509_V_ERR_UNSUPPORTED_NAME_SYNTAX;
+
+    memset(ulabel, 0, sizeof(ulabel));
+    /* Special case: initial '.' is RHS match */
+    if (*baseptr == '.') {
+        ulabel[0] = '.';
+        size -= 1;
+        if (a2ulabel(baseptr, ulabel + 1, &size) <= 0)
+            return X509_V_ERR_UNSPECIFIED;
+
+        if (size + 1 > eml->length) {
+            emlptr += size + 1 - base->length;
```

Comment:
This is still quite wrong.  You're trying to construct a suffix of emlptr of length `size +1`.  That's only possible if:
`eml->length >= size + 1`,  then the suffix pointer is `emlptr + (eml->length - size - 1)`.   If pointer arithmetic is too fragile for correct use, some macros or functions that are more difficult to misuse may be appropriate.  We cannot afford to make these types of mistakes. :-(

Reply:
Oops :(
Mea culpa.